src/thingsboard.c

```diff
@@ -189,6 +196,84 @@ static void client_request_time(struct k_work *work)
 	k_work_reschedule(k_work_delayable_from_work(work), K_SECONDS(10));
 }
 
+static int client_handle_rpc_response(struct coap_client_request *req, struct coap_packet *response)
+{
+	const uint8_t *payload;
+	uint16_t payload_len;
+	uint8_t code;
+	char code_str[5];
+	char expected_code_str[5];
+
+	LOG_INF("%s", __func__);
+
+	code = coap_header_get_code(response);
+	if (code != COAP_RESPONSE_CODE_CONTENT) {
+		coap_response_code_to_str(code, code_str);
+		coap_response_code_to_str(COAP_RESPONSE_CODE_CONTENT, expected_code_str);
+		LOG_ERR("Unexpected response code for RPC request: got %s, expected %s", code_str,
+			expected_code_str);
+		return -1;
+	}
+
+	payload = coap_packet_get_payload(response, &payload_len);
+	if (!payload_len) {
+		LOG_ERR("Received an empty RCP response");
+		return payload_len;
+	}
+
+	if (rpc_cb) {
+		rpc_cb(payload, payload_len);
+	}
+
+	k_sem_give(&rpc_sem);
+
+	return 0;
+}
+
+int thingsboard_rpc(const char *method, rpc_callback_t cb, ...)
+{
+	int err;
+	va_list param;
+	char params[RPC_PARAMS_SIZE];
+	char payload[RPC_PAYLOAD_SIZE];
+
+	if (method == NULL) {
+		LOG_ERR("method name must not be 'NULL'");
+		return -EINVAL;
+	}
+
+
+	va_start(param, cb);
+	bool params_exist = vsnprintf(params, sizeof(params), "%s", param) > 0 ? true : false;
+	if (!params_exist) {
+		strcpy(params, "{}");
+	}
+	va_end(param);
+
+	if (k_sem_take(&rpc_sem, K_MSEC(5000)) > 0) {
+		// in case no response was ever received, reset the semaphore and continue
+		k_sem_reset(&rpc_sem);
+		err = k_sem_take(&rpc_sem, K_MSEC(50));
+		if (err > 0) {
+			LOG_ERR("failed to take RPC semaphore");
+			return err;
+		}
+	}
```

Comment:
Not sure if the way you inplemented it in the send routine is the right way to do it.

What I assume you want to do is: If there is no response within 5 seconds of sending, then just assume, that there was never a response and send the next message.

What you achieve: When trying to send a message, wait for 5 seconds before actually sending it, so that the old message is gone.

and if two threads try to send at nearly the same time, then the second thread will overwrite the callback for the first thread.

Reply:
not sure I can follow. assuming there is no ongoing request: why would the code wait for 5 seconds? take would return immediately, as the semaphore is free. the second param to 'take' is a timeout, after all, isn't it?

Reply:
What I was hinting at was assume the following sequence of Events:
time in minute:seconds

```
0:00: call to thingsboard_rpc (1)
10:00: call to thingsboard_rpc (2)
10:04: call to thingsboard_rpc (3)
```

If we assume for 1 to not return, then 2 will wait 5 seconds for the semaphore, and at 10:05 it will reset the semaphore and take it.
Now it will create the request, overwrite the callback and return, making place for other threads to run.
3 will be scheduled to run, because the semaphore it was trying to take was reset, which in turn is an error on taking a semaphore, it will immediately reset the semaphore and take it, then send its own request overwriting the callback.

If we assume the response from 2 at 10:06, it is already to late, as the callback function has been overwritten by 3.

If you instead keep track on when you send out the request at the beginning, then you could reset the semaphore at 10:00, send your request, receive the answer at 10:01 and have no problems or conflicts with 3.